x-pack/docs/en/security/authorization/custom-authorization.asciidoc

```diff
@@ -0,0 +1,159 @@
+  `org.elasticsearch.xpack.core.security.SecurityExtension`. There you need to
+  override the following method:
++
+[source,java]
+----------------------------------------------------
+@Override
+public AuthorizationEngine getAuthorizationEngine(Settings settings) {
+    ...
+}
+----------------------------------------------------
++
+The `getAuthorizationEngine` method is used to provide the authorization engine
+implementation.
+
+Sample code that illustrates the structure and implementation of a custom
+authorization engine is provided in the
+https://github.com/elastic/elasticsearch/tree/master/plugin/examples/security-example-authorization-engine[elasticsearch]
+repository on GitHub. You can use this code as a starting point for creating your
+own authorization engine.
+
+[[packing-extension-plugin]]
+==== Implement an elasticsearch plugin
+
+In order to register the security extension, you need to also implement an
+elastiicsearch plugin that contains the extension:
+
+. Implement a plugin class that extends `org.elasticsearch.plugins.Plugin`
+. Create a build configuration file for the plugin; Gradle is our recommendation.
+. Create a `plugin-descriptor.properties` file as described in the
+  <<plugin-authors,plugin authors>> section.
+. Create a `META-INF/services/org.elasticsearch.xpack.core.security.SecurityExtension` descriptor file for the
+  extension that contains the fully qualified class name of your `org.elasticsearch.xpack.core.security.SecurityExtension` implementation
+. Bundle all in a single zip file.
+
+[[using-security-extension]]
+==== Using the security extension
+
+To use a security extension:
+
+. Install the plugin with the extension on each node in the cluster. You run
+  `bin/elasticsearch-plugin` with the `install` sub-command and specify the URL
+  pointing to the zip file that contains the extension. For example:
++
+[source,shell]
+----------------------------------------
+bin/elasticsearch-plugin install file:///<path>/my-extension-plugin-1.0.zip
+----------------------------------------
+
+. Add any configuration parameters for implementations in the extension to the
+`elasticsearch.yml` file.  The settings are not namespaced and you have access to any
+settings when constructing the extensions, although it is recommended to have a
+namespacing convention for extensions to keep your `elasticsearch.yml`
+configuration easy to understand.
++
+For example, if you have a custom roles provider that
+resolves roles from reading a blob in an S3 bucket on AWS, then you would specify settings
+in `elasticsearch.yml` such as:
++
+[source,js]
+----------------------------------------
+custom_roles_provider.s3_roles_provider.bucket: roles
+custom_roles_provider.s3_roles_provider.region: us-east-1
+custom_roles_provider.s3_roles_provider.secret_key: xxx
+custom_roles_provider.s3_roles_provider.access_key: xxx
+----------------------------------------
```

Comment:
Is this example a repetition of the one below on purpose? If there's a different example we could use, I think that would be more helpful.

Reply:
that was an oops